src/validation.cpp

```diff
@@ -834,7 +834,10 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
         }
     }
 
-    entry.reset(new CTxMemPoolEntry(ptx, ws.m_base_fees, nAcceptTime, m_active_chainstate.m_chain.Height(),
+    // Set entry_sequence to 0 when bypass_limits is used; this allows txs from a block
+    // reorg to be marked earlier than any child txs that were already in the mempool.
+    const uint64_t entry_sequence = bypass_limits ? 0 : m_pool.GetSequence();
```

Comment:
> Doesn't that mean there are races? (There is no guarantee that `TransactionAddedToMempool()` will always run in the near future)?

The mempool value is incremented when TrAddedTMP is queued, not when it's eventually run -- the increment happens first, then the result of the increment is passed through and queued up. So I don't think there's any racy here.

> Perhaps it makes sense to use `m_pool.GetSequence() + 1` here?

I think it would make more sense to change `info_for_relay`?

```c++
TxMempoolInfo CTxMemPool::info_for_relay(const GenTxid& gtxid, uint64_t last_sequence) const
{
    LOCK(cs);
    indexed_transaction_set::const_iterator i = (gtxid.IsWtxid() ? get_iter_from_wtxid(gtxid.GetHash()) : mapTx.find(gtxid.GetHash()));
//  if (i == mapTx.end() || last_sequence < i->GetSequence()) {
    if (i == mapTx.end() || last_sequence <= i->GetSequence()) {
        return TxMempoolInfo();
    } else {
        return GetInfo(i);
    }
}
```

Reply:
> Doesn't that mean there are races? (There is no guarantee that TransactionAddedToMempool() will always run in the near future)?

I meant `TransactionAddedToMempool` event happening, not the validation interface clients' callbacks. IIUC the event firing on main signals (which calls `GetAndIncrementSequence()`) happens in validation before `ProcessTransaction` returns:
https://github.com/bitcoin/bitcoin/blob/e5a9f2fb62dc4db6cad21b2997d96a881ea64125/src/validation.cpp#L1234

The queued callbacks are scheduled and may not have necessarily happened yet, but PeerManager doesn't implement `TransactionAddedToMempool`. It is only interested in the current `GetSequence()` and what is recorded in `mempool_sequence`.

> Basically, I am wondering what happens if you reorder t=2 and t=3 in your example.

My understanding is that `GetSequence()` returns 2 when `ProcessTransaction` is done, even if t=2 and t=3 are swapped.

Reply:
Ah I didn't see the last message.

> I think it would make more sense to change info_for_relay?

sounds good to me

Reply:
Updated. This also changes the initial value for `m_last_inv_sequence` to `1` instead of `0`. (Note that I reversed the order of the `if .. else` in `info_for_relay` so `GetSeq() < last` becomes `! (GetSeq() <= last)` which is `last < GetSeq()`)